src/common/buffer.cc

```diff
@@ -1065,29 +1065,30 @@ static ceph::spinlock debug_lock;
     ceph_assert(o+l <= _len);
     unsigned p = 0;
     for (auto& node : _buffers) {
-      if (p + node.length() > o) {
-        if (p >= o && p+node.length() <= o+l) {
-          // 'o'------------- l -----------|
-          //      'p'-- node.length() --|
-	  node.zero();
-        } else if (p >= o) {
-          // 'o'------------- l -----------|
-          //    'p'------- node.length() -------|
-	  node.zero(0, o+l-p);
-        } else if (p + node.length() <= o+l) {
-          //     'o'------------- l -----------|
-          // 'p'------- node.length() -------|
-	  node.zero(o-p, node.length()-(o-p));
-        } else {
-          //       'o'----------- l -----------|
-          // 'p'---------- node.length() ----------|
-          node.zero(o-p, l);
-        }
+      if (o + l <= p) {
+        break;
       }
-      p += node.length();
-      if (o+l <= p) {
-	break;  // done
+      if (node.length() == 0) {
+        continue;
       }
+      if (p >= o && p+node.length() <= o+l) {
+        // 'o'------------- l -----------|
+        //      'p'-- node.length() --|
+        node.zero();
+      } else if (p >= o) {
+        // 'o'------------- l -----------|
+        //    'p'------- node.length() -------|
+        node.zero(0, o+l-p);
+      } else if (p + node.length() <= o+l) {
+        //     'o'------------- l -----------|
+        // 'p'------- node.length() -------|
+        node.zero(o-p, node.length()-(o-p));
+      } else {
+        //       'o'----------- l -----------|
+        // 'p'---------- node.length() ----------|
+        node.zero(o-p, l);
```

Comment:
i think this is wrong. what if `p + node.length() <= o`?

Reply:
also "simplify" is way too vague, better off expanding the commit message with more details.

Reply:
@tchaikov 
You're right. we need to consider the case ````p + node.length() <= o``
I spend about 1 month on studying bufferlist design. There's no logic problem here. I just want to make it more clear when checking the code again.
Do I need to refine the change here or abandon it directly?

Reply:
personally, i think this loop is simple enough to digest with the help of ASCII art. if you can come up with a simpler and correct design. that'd be great.

Reply:
I'll abandon this change